### PR TITLE
[pytree] Use OpTree for PyTree manipulation

### PR DESCRIPTION
**Pull request by XuehaiPan**
Split from #92679. Use C++-based PyTree implementation.

## Highlights

1. High performance (20x speedup than the pure-Python implementation, 10%-20% overall speedup for `torch.fx`)
2. Multi-input tree-map support
3. Custom tree node registry with namespace isolation

Refs:

- #65761
- #91323
- #92679

From https://github.com/pytorch/pytorch/issues/65761#issuecomment-1334746366:

> ### 0. Out-of-box compatible with JAX's pytree, provides the same interfaces and functions (and more).
> 
> ### 1. High-performance: `optree` has comparable fast tree operations (~0.9x for `dict`s and ~2.5x for `OrderedDict`s) than JAX's pytree and it is 20x faster than `torch.utils._pytree`.
> 
> `optree` implements some common Python container types in C++ (e.g., `OrderedDict`) and achieves 2.5x performance than JAX's pytree. Check out section [Built-in PyTree Node Types](https://github.com/metaopt/optree#built-in-pytree-node-types) and [Benchmark](https://github.com/metaopt/optree#benchmark) for more details.
> 
> | Module    | Nodes | OpTree (μs) | JAX XLA (μs) | PyTorch (μs) | DM-Tree (μs) | Speedup (J / O) | Speedup (P / O) | Speedup (D / O) |
> | :-------- | ----: | ----------: | -----------: | -----------: | -----------: | --------------: | --------------: | --------------: |
> | TinyMLP   |    53 |       26.40 |        68.19 |       586.87 |        34.14 |            2.58 |           22.23 |            1.29 |
> | AlexNet   |   188 |       84.28 |       259.51 |      2182.07 |       125.12 |            3.08 |           25.89 |            1.48 |
> | ResNet18  |   698 |      288.57 |       807.27 |      7881.69 |       429.39 |            2.80 |           27.31 |            1.49 |
> | ResNet34  |  1242 |      580.75 |      1564.97 |     15082.84 |       819.02 |            2.69 |           25.97 |            1.41 |
> | ResNet50  |  1702 |      791.18 |      2081.17 |     20982.82 |      1104.62 |            2.63 |           26.52 |            1.40 |
> | ResNet101 |  3317 |     1603.93 |      3939.37 |     40382.14 |      2208.63 |            2.46 |           25.18 |            1.38 |
> | ResNet152 |  4932 |     2446.56 |      6267.98 |     56892.36 |      3139.17 |            2.56 |           23.25 |            1.28 |
> | ViT-H/14  |  3420 |     1681.48 |      4488.33 |     41703.16 |      2504.86 |            2.67 |           24.80 |            1.49 |
> | Swin-B    |  2881 |     1565.41 |      4091.10 |     34241.99 |      1936.75 |            2.61 |           21.87 |            1.24 |
> |           |       |             |              |              |  **Average** |        **2.68** |       **24.78** |        **1.38** |
> 
> <div align="center">
>   
> </div>
> 
> ### 2. Namespace Isolation for the PyTree Type Registry
>  
> In addition to the JAX's pytree registry for custom node type registration, `optree` adds `namespace` isolation to the registry. Users can register the same type multiple times for different flatten/unflatten behavior. It also provides module-level isolation for safety reasons. For example, you can add a unique prefix to your namespace to isolate your registry with other modules (e.g., `torch.xxx`, `torch.functorch.xxx`):
> 
> ```python
> # Register a Python type into a namespace
> import torch
> 
> optree.register_pytree_node(
>     torch.Tensor,
>     # (tensor) -> (children, metadata)
>     flatten_func=lambda tensor: (
>         (tensor.cpu().numpy(),),
>         dict(dtype=tensor.dtype, device=tensor.device, requires_grad=tensor.requires_grad),
>     ),
>     # (metadata, children) -> tensor
>     unflatten_func=lambda metadata, children: torch.tensor(children[0], **metadata),
>     namespace='torch.torch2numpy',
> )
> ```
> 
> ```python
> >>> tree = {'weight': torch.ones(size=(1, 2)).cuda(), 'bias': torch.zeros(size=(2,))}
> >>> tree
> {'weight': tensor([[1., 1.]], device='cuda:0'), 'bias': tensor([0., 0.])}
> 
> # Flatten without specifying the namespace
> >>> tree_flatten(tree)  # `torch.Tensor`s are leaf nodes
> ([tensor([0., 0.]), tensor([[1., 1.]], device='cuda:0')], PyTreeSpec({'bias': *, 'weight': *}))
> 
> # Flatten with the namespace
> >>> leaves, treespec = optree.tree_flatten(tree, namespace='torch.torch2numpy')
> >>> leaves, treespec
> (
>     [array([0., 0.], dtype=float32), array([[1., 1.]], dtype=float32)],
>     PyTreeSpec(
>         {
>             'bias': CustomTreeNode(Tensor[{'dtype': torch.float32, 'device': device(type='cpu'), 'requires_grad': False}], [*]),
>             'weight': CustomTreeNode(Tensor[{'dtype': torch.float32, 'device': device(type='cuda', index=0), 'requires_grad': False}], [*])
>         },
>         namespace='torch.torch2numpy'
>     )
> )
> 
> # `entries` are not defined and use `range(len(children))`
> >>> optree.tree_paths(tree, namespace='torch.torch2numpy')
> [('bias', 0), ('weight', 0)]
> 
> # Unflatten back to a copy of the original object
> >>> optree.tree_unflatten(treespec, leaves)
> {'bias': tensor([0., 0.]), 'weight': tensor([[1., 1.]], device='cuda:0')}
> ```
> 
> Check out section [Registering a Container-like Custom Type as Non-leaf Nodes](https://github.com/metaopt/optree#notes-about-the-pytree-type-registry) for more details.
> 
> ### 3. Support both `None` as Non-leaf Node and `None` as Leaf
> 
> In JAX's implementation, `None` is always an internal non-leaf node with an arity 0, which is like an empty tuple. This limits the usage of the JAX's pytree utilities for PyTorch. For example, the `nn.Module` uses `_parameters` and `_buffers` (`OrderedDict[str, Optional[Tensor]]`) to hold the tensors, while the value can be a tensor or `None`.
> 
> `optree` supports both `None` as Non-leaf Node (JAX's default) and `None` as Leaf (PyTorch's default). Check out section [None is Non-leaf Node vs. None is Leaf](https://github.com/metaopt/optree#none-is-non-leaf-node-vs-none-is-leaf) for more details.
> 
> ### 4. Some other improvements and bug fixes
> 
> 1. Adds in-place version of treemap (`tree_map_`), which reduces redundant unflatten operation for better performance.
> 2. Adds support for tree flatten and tree map with paths. (useful for `functorch` module extraction).
> 3. Improves the JAX's pytree sorting support for `dict`s.
> 4. Better string representation `repr(PyTreeSpec)`.
> 5. Fixes some bugs for JAX's pytree of hashing, pickle serialization, segmentation fault for infinite recursion, and tree-compose/tree-transpose.

From https://github.com/pytorch/pytorch/pull/92679#issuecomment-1398778481:

> ```python
> # pytree_make_fx_bench.py
> import torch
> from torch.fx.experimental.proxy_tensor import make_fx
> import time
> 
> def f(x):
>     for _ in range(10000):
>         x = x+x
>     return x
> 
> import time
> begin = time.time()
> out = make_fx(f, tracing_mode="real")(torch.randn(20))
> begin = time.time()
> print(f'tracing_mode="real" {time.time() - begin:.2f}')
> out = make_fx(f, tracing_mode="fake")(torch.randn(20))
> print(f'tracing_mode="fake" {time.time() - begin:.2f}')
> 
> out = make_fx(f, tracing_mode="symbolic")(torch.randn(20))
> print(f'tracing_mode="symbolic" {time.time() - begin:.2f}')
> ```
> 
> This seems to run around 10-20% faster with the optree implementation:
> 
> ```
> # Optree
> python pytree_make_fx_bench.py
> tracing_mode="real" 0.00
> tracing_mode="fake" 6.32
> tracing_mode="symbolic" 27.13
> ```
> 
> ```
> # torch.utils._pytree
> python pytree_make_fx_bench.py
> tracing_mode="real" 0.00
> tracing_mode="fake" 7.66
> tracing_mode="symbolic" 31.07
> ```

cc @zou3519 @soumith

### CODE REVIEW

**Comment by XuehaiPan** [comment]
Seems that the Windows test runners do not install the requirements correctly.

File: pyproject.toml
Comment (Skylion007) [inline pyproject.toml:14]: Is this actually needed as a build-requirement? I don't think so, it can be installed as a project requirement.

File: torch/utils/pytree.py
Comment (Skylion007) [inline torch/utils/pytree.py:62]:
```suggestion
    FlattenFunc = Callable[[PyTree], Tuple[List, Optional[Context]]]
```
based on your other file? These mypy type notations aren't consistent with usage in torch.fx

File: torch/fx/pytree.py
Comment (Skylion007) [inline torch/fx/pytree.py:18]: Maybe the second arg here should be Optional too?

File: pyproject.toml
Comment (Skylion007) [reply pyproject.toml:14]: Yeah it should be installed in setup.py, but it doesn't need to be in the isolated build env

**Comment by XuehaiPan** [comment]
Requested `optree` on anaconda default channel in ContinuumIO/anaconda-issues#13131.

**Comment by Skylion007** [comment]
> Requested `optree` on anaconda default channel in [ContinuumIO/anaconda-issues#13131](https://github.com/ContinuumIO/anaconda-issues/issues/13131).
> 
> Is it possible to use the PyPI package rather than install it from the conda default channel? Or install it from the conda-forge channel? ([Conda Version](https://anaconda.org/conda-forge/optree) / [Conda Platforms](https://anaconda.org/conda-forge/optree))

What about the PyTorch channel, although I don't know who controls that? @malfet? @soumith?

**Comment by Skylion007** [comment]
@malfet might be the right person to ask about this issue?

**Comment by malfet** [comment]
@XuehaiPan please extended PR description that highlights the benefits of optree? Some benchmark perhaps... Also, lack of Anaconda package feels like a serious hindrance.

if this is a performance only feature, can it be an optional dependency?

**Comment by XuehaiPan** [comment]
> please extended PR description that highlights the benefits of optree? Some benchmark perhaps...

@malfet I update the PR description from some referenced discussion. As for benchmark, the C++-based PyTree utilities gain 20x speedup than the pure-Python implementation and achieve 10%-20% overall speedup for `torch.fx`.

> Also, lack of Anaconda package feels like a serious hindrance.

There is a conda-forge package with all platform and arch support. I have request the `optree` package on the default channel in ContinuumIO/anaconda-issues#13131. Waiting for the responses from the Anaconda team.

> if this is a performance only feature, can it be an optional dependency?

I think it's a performance-only feature as of now. But `optree` provides more features, e.g., registry namespace isolation, which may be used by some parts of `torch` in the future (e.g., `torch.distributed` https://github.com/pytorch/pytorch/pull/93214#discussion_r1090010384). I'm afraid it cannot be optional. Because some submodules directly import `torch.utils._pytree` (will migrate to `torch.utils.pytree` in the future which imports `optree`). For example, `torch.return_types` is a top-level submodule, its source code will be executed when users use `import torch`.

https://github.com/pytorch/pytorch/blob/4d51c8532c463831453835f4f49270d5bae85d1a/torch/return_types.py#L1-L34

**Comment by Chillee** [comment]
@malfet I think it's worth bringing in optree as a dependency inside PyTorch - we use pytrees all over the place and it's ended up being a fairly significant perf bottleneck during tracing. We've benchmarked about a 15-20% speedup in tracing by simply switching to optree.

I think my main uncertainty on this is what the... right way to build this is, both for the OSS release as well as internally within Meta. Since it requires an abseil dependency, which from my understanding is an issue within Meta.

**Comment by malfet** [comment]
@Chillee imo considering the size of the project, I wonder if @XuehaiPan would be willing to propose a PR that just adds entire source code and build scripts to the `pytorch/pytorch` repo? This way, we would not have problems with pybind11 version mismatch, platform availability (like anaconda, aarch64, upcoming windows  arm64, you-name-it) as well as synchronizing it with internal build system, etc.

**Comment by Skylion007** [comment]
> @Chillee imo considering the size of the project, I wonder if @XuehaiPan would be willing to propose a PR that just adds entire source code and build scripts to the `pytorch/pytorch` repo? This way, we would not have problems with pybind11 version mismatch, platform availability (like anaconda, aarch64, upcoming windows arm64, you-name-it) as well as synchronizing it with internal build system, etc.

Hmm, would pybind11 version mismatch be that much of an issue here? Only if you plan to cast back to optree in C++ code: The pybind11 version mismatch is only an issue if you are planning on casting those objects back into pytorch's C++.

I looked at the abseil code in the repo btw, and it mostly relates to str formatting, flat_hashmap and hash function, and std::span backports.

**Comment by Skylion007** [comment]
Could always just be added a submodule though I suppose if we don't plan on using it all outside of PyTorch.

**Comment by malfet** [comment]
> Hmm, would pybind11 version mismatch be that much of an issue here? Only if you plan to cast back to optree in C++ code: The pybind11 version mismatch is only an issue if you are planning on casting those objects back into pytorch's C++.

Not specific to a pybind, but on many operating systems there is a global namespace for symbols, so if library is not compiled with `-fvisibility-hidden` (on Linux) or if symbol is marked as global, only one instance of symbol will be used. So if one have `std::vector<int>::push_back` implementation in `libtorch_cpu.so` and `optree.so` only one implementation will be used, which if they compiled assuming different structure layout, can lead to a very weird crashes.  I've encountered a few with different versions of pybind.

**Comment by XuehaiPan** [comment]
> Not specific to a pybind, but on many operating systems there is a global namespace for symbols, so if library is not compiled with `-fvisibility-hidden` (on Linux) or if symbol is marked as global, only one instance of symbol will be used.

`optree` is built with `-fvisibility-hidden` flag:

```cmake
set(CMAKE_POSITION_INDEPENDENT_CODE ON)  # -fPIC
set(CMAKE_CXX_VISIBILITY_PRESET hidden)  # -fvisibility=hidden
```

**Comment by XuehaiPan** [comment]
I think this PR is ready except for the conda package from the anaconda default channel (requested in ContinuumIO/anaconda-issues#13131). Note that `optree` is available on [PyPI (Win / Linux / macOS for both x86_64 and ARM64)](https://pypi.org/project/optree/#files) and [conda-forge](https://anaconda.org/conda-forge/optree).

File: .github/workflows/_win-test.yml
Comment (XuehaiPan) [inline .github/workflows/_win-test.yml:128]: I add this for the Windows CI because the dependency in `setup.py` is not installed correctly.

Comment (voznesenskym) [reply .github/workflows/_win-test.yml:128]: cc @malfet

**Comment by voznesenskym** [comment]
@XuehaiPan as always, please lmk how I can help.

**Comment by XuehaiPan** [comment]
I think the blocker is the anaconda package on the default channel (ContinuumIO/anaconda-issues#13131). I wonder if we could provide `optree` first in the pytorch channel? cc @malfet

Ref https://github.com/conda-forge/optree-feedstock

**Comment by seemethere** [comment]
> I think the blocker is the anaconda package on the default channel ([ContinuumIO/anaconda-issues#13131](https://github.com/ContinuumIO/anaconda-issues/issues/13131)). I wonder if we could provide `optree` first in the pytorch channel? cc @malfet
> 
> Ref [conda-forge/optree-feedstock](https://github.com/conda-forge/optree-feedstock)

This is something we could do. cc @atalman for help with this.

File: torch/utils/pytree.py
Comment (Skylion007) [inline torch/utils/pytree.py:258]: Nit: this should be a TypeError not a ValueError, right?

Comment (XuehaiPan) [reply torch/utils/pytree.py:258]: This is copied from `_pytree.py`.

Comment (Skylion007) [reply torch/utils/pytree.py:258]: Yeah, I've been meaning do a sweep of the codebase with tryceratops https://github.com/guilatrova/tryceratops to flag and fix these.

File: .github/workflows/_win-test.yml
Comment (malfet) [reply .github/workflows/_win-test.yml:154]: cc: @ZainRizvi would it be possible to move Windows CI a few steps closer to other OSes by using wheel installation?

**Review by seemethere — CHANGES_REQUESTED** [review]
Given that `pytree` seems to be limited to `fx` for now does it make sense to make `optree` an optional dependency? Similar to what we currently do with `numpy`?

cc @malfet since you were the one who implemented our `numpy` optional dependency

File: .github/workflows/_win-test.yml
Comment (ZainRizvi) [reply .github/workflows/_win-test.yml:154]: Sounds reasonable, not sure why we don't do that already. 

cc @huydhn in case you know why we don't know wheels for Windows CI today

**Comment by XuehaiPan** [comment]
@seemethere

> Given that `pytree` seems to be limited to `fx` for now

This is not true (distributed, dynamo, functorch, fx, inductor, prims). There are many direct imports to `torch.utils._pytree`, which will migrate to the C++ implementation `torch.utils.pytree` in follow-up PRs. In addition, custom pytree type registration needs to execute in the global scope:

https://github.com/pytorch/pytorch/blob/3faa6361969e8cae35cb32c026df3a7d71b4f5ef/torch/return_types.py#L9-L34

https://github.com/pytorch/pytorch/blob/3faa6361969e8cae35cb32c026df3a7d71b4f5ef/torch/fx/immutable_collections.py#L51-L52

This needs a successful import for `torch.utils.pytree`.

------

> does it make sense to make `optree` an optional dependency? Similar to what we currently do with `numpy`?

I have made `optree` out-of-box support `PyStructSequence` types.

- #75982
- metaopt/optree#29
- metaopt/optree#30
 
So the global reference to `torch.utils.pytree` in `torch.return_types` can be removed. I think we could have a try at this. Other global imports:

```python
from torch.utils.pytree import tree_map

def somefunc():
    ret = tree_map(fn, tree)
```

can be replaced to:

```python
import torch

def somefunc():
    ret = torch.utils.tree_map(fn, tree)
```

I made `optree` optional in this PR. But I'm not sure it could be in follow-up PRs in the migration.

**Comment by voznesenskym** [comment]
> Given that `pytree` seems to be limited to `fx` for now does it make sense to make `optree` an optional dependency? Similar to what we currently do with `numpy`?
> 
> cc @malfet since you were the one who implemented our `numpy` optional dependency

I think we should make optree a true dep.

File: .github/workflows/_win-test.yml
Comment (huydhn) [reply .github/workflows/_win-test.yml:154]: I'm not sure about the historical context, but we didn't build wheel for Windows in the past.  We just installed PyTorch instead `python setup.py install --cmake`, thus the use of 7z-ing.  The good news is that this has already been change by https://github.com/pytorch/pytorch/commit/3cdf621fe5c8f8378fda209b8a143443d33b2086#diff-4e1dd3a0537c1f0f7bc80a030a72c818c4a1aae8c5c2fd62752e1c6b1d1d6288, in which the wheel is now built by `python setup.py bdist_wheel`.  The 7z part remains because changing that is not the intention of the PR.  Let me build on that and submit a change to use the built wheel instead

**Comment by Chillee** [comment]
I talked with @malfet about this, and there's 2 easy short-term paths to getting this landed.

1. Make it an optional dependency, and have a fallback to the python pytree when optree isn't installed. This is what you've done in this PR. Note that this does *not* preclude us from packaging optree by default on Linux. However, @malfet points out that making optree a "required" dependency forces all platforms (windows, arm, weirder stuff, etc.) that Pytorch supports to also figure out how to build optree. It may be worth doing that down the line, but that's a pretty significant lift now.

2. We vendor optree into the PyTorch repo. Basically, copy-paste the code into PyTorch, and then we become responsible for building optree inside of PyTorch. In some sense, this is similar to making optree a "required" dependency, except that PyTorch developers take ownership of making sure it builds.

Given that we've made optree an optional dependency in this PR, anything else blocking us from landing this? cc: @malfet

**Comment by Skylion007** [comment]
> abseil dependency

New protobuf versions will also require abseil FYI so it's going to get into the codebase at some point.

File: .github/workflows/_win-test.yml
Comment (XuehaiPan) [reply .github/workflows/_win-test.yml]: Removed since #94958 is landed.

- #94958

**Comment by zou3519** [comment]
Just to poke at this a bit more -- @XuehaiPan, are there still things to do here, or are we waiting on pytorch-dev-infra to give their approval?

**Comment by XuehaiPan** [comment]
> are there still things to do here, or are we waiting on pytorch-dev-infra to give their approval?

@zou3519 I think we are waiting for the anaconda package (ContinuumIO/anaconda-issues#13131) and then approval from pytorch-dev-infra. Since `optree` is pip / conda-forge installable and is made optional in this PR, I think pytorch-dev-infra folks can review this somehow without an anaconda package.

**Comment by voznesenskym** [comment]
Can I commandeer this

File: .github/workflows/_win-test.yml
Comment (malfet) [inline .github/workflows/_win-test.yml:165]: Why this is necessary, as we don't do any `torch.compile` testing on Windows?

**Review by malfet — COMMENTED** [review]
I'm confused what this PR is trying to do at this point. It introduces few new classes an optional dependency on `optree`, but they don't seem to be used/tested anywhere

**Comment by XuehaiPan** [comment]
> I'm confused what this PR is trying to do at this point. It introduces few new classes an optional dependency on `optree`, but they don't seem to be used/tested anywhere

Copied from https://github.com/pytorch/pytorch/pull/92679#issuecomment-1399340200:

> Diff 1 add new dep
> Diff 2 Integration behind disabled flag
> Diff 3 pass tests with flag enabled and disabled
> Diff 4 set flag to default on
> let soak a few weeks
> Diff 5 remove flag, remove old system

@malfet The ultimate goal is to replace the slow pure-Python PyTree implementation with the CXX implementation. This PR is the first step to introduce the `optree` package as an optional dependency. The migration process will begin after this PR has landed.

I will add tests for the `optree`-based PyTree in this PR. Since `optree` is only an optional dependency so far and not imported, these tests will only run on our CI and not on the user side.

File: .github/workflows/_win-test.yml
Comment (XuehaiPan) [reply .github/workflows/_win-test.yml:165]: The final goal for this migration is to replace all pure-Python-based PyTree implementation (`torch.utils._pytree`), which is already used in some downstream packages (e.g., `torchvision`). I will add tests for `optree` in some follow-up commits. See also my comment https://github.com/pytorch/pytorch/pull/93139#issuecomment-1722141585.

**Comment by XuehaiPan** [comment]
I have added some tests for the CXX-based PyTree implementation.

I also opened an issue to discuss how to serialize/deserialize the treespec object in:

- #109460

File: torch/fx/pytree.py
Comment (Chillee) [inline torch/fx/pytree.py:1]: Why do these changes need to be made in this diff?

File: test/test_pytree.py
Comment (Chillee) [inline test/test_pytree.py:249]: Are most of these just formatting changes?

File: torch/fx/pytree.py
Comment (XuehaiPan) [reply torch/fx/pytree.py:1]: If this is unrelated, I can remove this file change.

This file only simulates the behavior of `torch.fx._pytree` using the new `torch.utils.pytree`.

- `torch.utils.pytree` <-> `torch.utils._pytree`
- `torch.fx.pytree` <-> `torch.fx._pytree`

Comment (Chillee) [reply torch/fx/pytree.py:1]: I understand that, but my understanding was that the plan was to isolate this PR as "only adding dependencies", and then we can incrementally port all uses of pytrees over in subsequent PRs.

File: test/test_pytree.py
Comment (XuehaiPan) [reply test/test_pytree.py:249]: Yes.

File: torch/fx/pytree.py
Comment (XuehaiPan) [reply torch/fx/pytree.py:1]: Get it.

File: test/test_pytree.py
Comment (XuehaiPan) [reply test/test_pytree.py:249]: Note that it also fixes some tests for `tree_map` and serialization. All extra test changes have passed in the CI keeping the `torch.utils._pytree` untouched. I can also split these test changes in a separate PR if requested. One for test fixes for the Python-based PyTree and one for introducing the new dependency (this PR).

Comment (XuehaiPan) [reply test/test_pytree.py:249]: Opened a PR for better code review:

- #109483

**Review by malfet — APPROVED** [review]

File: torch/utils/pytree.py
Comment (Chillee) [inline torch/utils/pytree.py:30]: Should we be guarding this import with a try/catch and fallback to the python pytree if the dependency doesn't exist?

Comment (XuehaiPan) [reply torch/utils/pytree.py:30]: `torch.utils.pytree` is not imported when we write `import torch`. Maybe we can guard this like we do for `numpy` and prompt a warning.

Comment (Chillee) [reply torch/utils/pytree.py:30]: The issue is the next steps. i imagine the next steps are something like "incrementally replace all uses of `torch.utils._pytree` with `torch.utils.pytree`", right?

Right now this isn't a problem, but once a file actually starts using `torch.utils.pytree`, this is an issue. I think we need to guard on `optree` being present and if it isn't, fall back to using `torch.utils._pytree`.

Comment (XuehaiPan) [reply torch/utils/pytree.py:30]: I'll submit a follow-up PR to resolve this.

Comment (Chillee) [reply torch/utils/pytree.py:30]: Thanks @XuehaiPan! I'm excited to start porting PyTorch to use optree :) It's been a long time delayed.

**Pull request closed without merging**
